### PR TITLE
Add functionality to shift the entire map

### PR DESCRIPTION
**Pull request by IntelOrca**
Allows increasing the map size at 0, 0 direction.
New parameters are added to the change map size game action so that the map is shifted after increasing the map size.

Needs testing to make sure all stored coordinates across tiles, entities, maps, banners, everything are shifted correctly.

Zip contains a saved game for testing and a plugin to trigger the action with shift 8, 8
[shiftmap.zip](https://github.com/OpenRCT2/OpenRCT2/files/11524752/shiftmap.zip)

### CODE REVIEW

File: src/openrct2/world/Map.cpp [hunk 1]
```diff
@@ -2389,6 +2390,10 @@ void ShiftMap(const TileCoordsXY& amount)
         for (const auto& entityId : list)
         {
             auto entity = GetEntity(entityId);
+
+            // Do not tween the entity
+            EntityTweener::Get().RemoveEntity(entity);
```
Comment (ZehMatt) [inline src/openrct2/world/Map.cpp:2395]: Probably best to get a reference to the tweener outside of the loop.

**Review by ZehMatt — APPROVED** [review]
Needs a rebase and I tested this a while back already, not sure why I haven't approved it yet, sorry about that.

**Comment by fidwell** [comment]
I tested this with a more complex map, and noticed that something weird happens with the 3D sign scenery. If you do a shift with the plugin, signs appear to shift as normal. But if you click on them, the window that pops up has a blank viewport. Then, if you save, close, and re-open the map, the signs are blank.

**Comment by IntelOrca** [comment]
> I tested this with a more complex map, and noticed that something weird happens with the 3D sign scenery. If you do a shift with the plugin, signs appear to shift as normal. But if you click on them, the window that pops up has a blank viewport. Then, if you save, close, and re-open the map, the signs are blank.

Nice find, should now fixed.

**Comment by fidwell** [comment]
Yeah, that seems to have done the trick, thanks! Excited for this feature go out!

File: src/openrct2/world/Map.cpp [hunk 2]
```diff
@@ -2227,3 +2232,176 @@ MapRange ClampRangeWithinMap(const MapRange& range)
     MapRange validRange = MapRange{ aX, aY, bX, bY };
     return validRange;
 }
+
+void ShiftMap(const TileCoordsXY& amount)
+{
+    if (amount.x == 0 || amount.y == 0)
+        return;
```
Comment (Broxzier) [inline src/openrct2/world/Map.cpp:2239]: Shouldn't this be using `&&`? Now if either x or y is zero, it won't shift anything.

File: src/openrct2/world/Map.cpp [hunk 3]
```diff
@@ -2227,3 +2232,176 @@ MapRange ClampRangeWithinMap(const MapRange& range)
     MapRange validRange = MapRange{ aX, aY, bX, bY };
     return validRange;
 }
+
+void ShiftMap(const TileCoordsXY& amount)
+{
+    if (amount.x == 0 || amount.y == 0)
```
Comment (IntelOrca) [inline src/openrct2/world/Map.cpp:2238]:
```suggestion
    if (amount.x == 0 && amount.y == 0)
```

**Review by ZehMatt — DISMISSED** [review]
I went over the code again and noticed that jumping fountain entities aren't moved, I'm also not sure if the map animations need to be re-created for when its moved.
Edit: After saving and reloading the park the animations are working again so I'm guessing the animations need to be fixed when shifted.

**Comment by fidwell** [comment]
Keep open please!

**Comment by fidwell** [comment]
Please keep open!!

**Comment by Gymnasiast** [comment]
I rebased this PR and added shifting map animations and jumping fountains. It needs some more testing.

**Comment by ZeeMaji** [comment]
When testing with the included park on the merry go round, as well as BPB with the dodgems. for some reason it thinks the entrance/exit aren't connected when you shift the map? They still function fine but it spams the news bar with unconnected entrances/exits warnings.
Screenshot_select-area_20240330210355

**Comment by Basssiiie** [comment]
Similarly, peeps don't seem to be interacting with shifted stalls anymore.

image

**Comment by astraylife1** [comment]
Trying to test this but I'm not sure how to get the map shifting to function. I have the plugin installed but doesn't appear anything is happening. I might be missing a step

**Comment by Gymnasiast** [comment]
> Trying to test this but I'm not sure how to get the map shifting to function. I have the plugin installed but doesn't appear anything is happening. I might be missing a step
> 
> **edit, does it only work on the test park?

The plugin adds an option under the Map icon called "shift map". When you click that, it should shift by 8 tiles on both directions.

File: src/openrct2/actions/MapChangeSizeAction.cpp [hunk 4]
```diff
@@ -63,6 +70,9 @@ GameActions::Result MapChangeSizeAction::Execute() const
         MapExtendBoundarySurfaceY();
     }
 
+    // Shift the map (allows increasing the map at the 0,0 position
+    ShiftMap(_shift);
```
Comment (Gymnasiast) [inline src/openrct2/actions/MapChangeSizeAction.cpp:74]: Note: check if it correctly handles shifting a (say) 999×999 map by 1×1, or a 995×995 map by 8×8 (which you cannot do).

**Comment by astraylife1** [comment]
I found an issue with rides with an entrance and exit station. Normally if the ride is set to only dispatch trains with full loads, the train will leave the exit only station regardless, since the game knows it is not possible to obtain a full load. Once the map is shifted, the trains will back up at the exit station waiting for peeps.

As far as maps at 992x992 etc, the game won't let you shift since it would go above the limit. 

Peeps don't interact with stalls after shift

Every park I have shifted seems to have one ride where you continually get the warning that guests cannot reach entrance and no path leading from exit. I cannot find a pattern of how the ride is selected but it's always one ride and it seems to be the same ride if I repeat the experiment.

**Comment by Gymnasiast** [comment]
@ZeeMaji @Basssiiie @astraylife1 The cause seems to have been accidentally shifting coordinates that were set to NULL, meaning the game now expected entrances and exits in places where it didn’t previously. I just pushed an update to fix that, could you retest?

**Comment by ZeeMaji** [comment]
Seems to be working fine now, but i've also just now noticed that guest spawners don't seem to function properly after being shifted if they're shifted off of an edge tile? They can still spawn guests as normal, but guests can no longer despawn from this guest spawner as they will just walk to the edge of the path and turn back around instead of despawning. Not sure if this is just an issue with guest spawners themselves or something specific to shifting them in here.

Does not affect spawners which remain connected to the edge.

Screenshot_select-area_20240424155802

**Comment by astraylife1** [comment]
It was my understanding that guest spawners only spawn guests. They don’t despawn. Path on the edge of the map is how guests despawn unless I am mistaken.

**Comment by ZeeMaji** [comment]
Right, i recall it being something like that now that you mention it. Not sure if that's something that should be changed before this is implemented or if it's fine to disregard it as something separate for now.

**Comment by astraylife1** [comment]
I wouldn’t change it, as that is vanilla behavior. The spawn points already don’t move when you enlarge the map in sandbox mode.

**Comment by Gymnasiast** [comment]
I checked the code to be sure. When a guest that isn’t in the park heads into the map edge, a flag is set. All places that check this flag remove the guest if it is. So @astraylife1 ’s understanding is correct.

----------------------------

Just to be sure, I’d like @Basssiiie to also confirm everything now works. Then I can bump the network version and get it merged.

**Comment by astraylife1** [comment]
Will this function be added to the game via the plugin that was created or become an actual feature built in? Either way, I feel that there should be a prompt asking to confirm that shifting the map is what the user wants, as there is no way to shift it back afterward. And it is easy to accidentally select shift map in the plugin drop down instead of selecting scenery manager for example.

**Comment by Gymnasiast** [comment]
I haven’t really thought about the interface. The plugin is just a quick-and-dirty piece of code for testing though - as interfaces go, the absolute minimum would be to ask the user to specify what to shift by.

**Comment by Basssiiie** [comment]
One further issue I found is that the location of "Overall view" in ride windows don't shift with the ride/stall during shifting until the window is reopened, but IMHO it's pretty minor.

image

And another one is that shifting the map leaves around some unassigned patrol areas that are not linked to a staff member anymore (I'm guessing).

image

**Comment by Gymnasiast** [comment]
@Basssiiie Fixed the patrol areas bug. Looked into the viewports as well, but that is a bit too much faff.

Could you test it again? Hope we can ship this for v0.4.12.

**Review by Basssiiie — APPROVED** [review]
I think we got 'em all. I can't find any further issues, nice work! 😄

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**